README.md
# node-xen-backup

[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)
[![NPM version](https://img.shields.io/npm/v/node-xen-backup.svg)](https://www.npmjs.com/package/node-xen-backup) 

<b>node-xen-backup</b> is a Node.js module that uses xen-api to take a snapshot of a Xenserver VM and export to a file for backup. The module includes a cli to execute direct on bash.

## You can contribute

It is a free software and its use is free for everyone, but if you want to contribute anyway, please submit a donation.

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=G9AL2VHXZRL92&lc=US&item_name=Xenserver%20backup%20script%20development%20costs%20and%20efforts&item_number=0001&currency_code=USD&bn=PP%2dDonationsBF%3abtn_donateCC_LG%2egif%3aNonHosted)
 
## Support

 * Node.js (tested on >= 7.x)
 
## Installation

### Installing global for cli usage
```sh
> npm install -g node-xen-backup
```

### Installing as node module
```sh
> npm install node-xen-backup
```
```javascript
  const XenBackup = require('node-xen-backup')
```

## Cli

### Getting help

 ```sh
> xen-backup --help

# Usage: xen-backup [options]


# Options:
#
#    -V, --version              output the version number
#    -H, --host [host]          IP or hostname from Xenserver. Default: localhost.
#    -P, --port [port]          TCP port from Xenserver. Default: 80.
#    -u, --username [username]  Username to login in Xenserver.
#    -p, --password <password>  Password to login in Xenserver.
#    -f, --filename <filename>  Path to target file of exported snapshot.
#    -t, --type <type>          Type of object to export. <vm|vdi> default vm.
#    -v, --vm <vm>              VM name label or uuid to backup.
#    -h, --help                 output usage information
 ```

### Creating a backup to a VM
```sh
# Backup command
> xen-backup -H 192.168.1.1 -P 80 -u root -p password -f /backup/VMLinux.xva -v VMLinux
# Normal output
- Searching for VM VMLinux
- Searching VM real reference for UUID 38d15c62-a1a2-f5f8-de3d-55fb360e96c8
- Taking snapshot...
- Getting snapshot data...
- Exporting snapshot VMLinux1516035235 taked in January 15, 2018
- Will be saved on file /backup/VMLinux.xva
- The file has size 784.93MB
- Removing snapshot...
- Snapshot removed successfuly
```

## API

### Creating a backup
Example creating a backup using as Node.js module. This is very much like procedure used in cli.

```javascript
// [another requires...]

const XenBackup = require('node-xen-backup')

const connOptions = {
  username: 'root',
  password: 'password',
  host: '192.168.1.1',
  port: 80,
  sessionId: null
}
const destFileName = "/backup/filename.xva"

// First sets connecion options.
XenBackup.setClient(connOptions.host, connOptions.port)
// Login in xenserver xen-api.
XenBackup.login(connOptions.username, connOptions.password, (err, sessionId) => {
  if (err) return console.error(err)
  // Once logged sets session ID
  connOptions.sessionId = sessionId
  // Calls function to get itens
  getItens((err, itens) => {
    if (err) return console.error(err)
    let founded = false
    // Search for VM or VDI in returned itens
    for (let i in itens) {
      if (itens[i].uuid.toLowerCase() === program.vm.toLowerCase() || itens[i].name.toLowerCase() === program.vm.toLowerCase()) {
        founded = true
        // Takes snapshot
        itens[i].snapshot(itens[i].name.replace(/\s/g, '') + moment().unix(), (err, snapshot) => {
          if (err) return console.error(err)
          let protocol = 'http'
          if ((connOptions.port === '443') || (connOptions.port === 443)) protocol = 'https'
          // Call function to export snapshot
          downloadBackup(protocol, snapshot.uuid, (err, download) => {
            if (err) console.error(err)
            // Removes snapshot
            snapshot.remove((err, removed) => {
              if (err) return console.error(err)
              // SUCCESS!!!
              return console.log("VM OR VDI EXPORTED")
            })
          })
        })
      }
    }
    if (!founded) {
      return console.error('VDI or VHD not found in Xenserver.')
    }
  })
})

function getItens (callback) {
  if (program.type === 'vm') {
    XenBackup.getAllVMs((err, vms) => {
      if (err) return callback(err, null)
      return callback(null, vms)
    })
  }
  if (program.type === 'vdi') {
    XenBackup.getAllVDIs((err, vdis) => {
      if (err) return callback(err, null)
      return callback(null, vdis)
    })
  }
  if ((program.type !== 'vdi') && (program.type !== 'mv')) {
    return callback(new Error('Cannot define object type <vm|vdi>.'))
  }
}

function downloadBackup (protocol, uuid, callback) {
  let sufix = null
  if (program.type === 'vm') {
    sufix = `/export?session_id=${connOptions.sessionId}&use_compression=true&uuid=`
  }
  if (program.type === 'vdi') {
    sufix = `/export_raw_vdi?session_id=${connOptions.sessionId}&format=vhd&vdi=`
  }
  let req = request.get(`${protocol}://${connOptions.host}:${connOptions.port}${sufix}${uuid}`)
    .on('end', function (response) {
      return callback(null, response)
    })
    .on('error', function (err) {
      console.log(err)
      return callback(err, null)
    })
  if (program.type === 'vdi') {
    req.pipe(GZip).pipe(fs.createWriteStream(destFileName))
  } else {
    req.pipe(fs.createWriteStream(destFileName))
  }
}
```


CHANGELOG.md
# v0.3.0
- Methods for API usage simplified.
- Support for VDI export.
# v0.2.3
- API documentation improved on README.
- Request to export URL on xenserver now uses https when port is set to 443.
- Node process now ignores tls certificate verification.


lib/vm.js
function VM (data, client, sessionId) {
  this.uuid = null
  this.ref = null
  this.name = null
  this.isTemplate = null
  this.isSnapshot = null
  this.isControlDomain = null
  this.powerState = null
  this.snapshotTime = null
  this.client = null
  this.sessionId = null
  if (typeof sessionId !== 'undefined') this.sessionId = sessionId
  if (typeof data.uuid !== 'undefined') this.uuid = data.uuid
  if (typeof data.ref !== 'undefined') this.ref = data.ref
  if (data.name || data.name_label) this.name = data.name || data.name_label
  if (typeof data.is_a_template !== 'undefined') this.isTemplate = data.is_a_template
  if (typeof data.is_a_snapshot !== 'undefined') this.isSnapshot = data.is_a_snapshot
  if (typeof data.is_control_domain !== 'undefined') this.isControlDomain = data.is_control_domain
  if (typeof data.power_state !== 'undefined') this.powerState = data.power_state.toLowerCase()
  if (typeof data.snapshot_time !== 'undefined') this.snapshotTime = data.snapshot_time
  // GETS VM REFS
  if (typeof client !== 'undefined') this.client = client
}

VM.prototype.snapshot = function (snapshotName, callback) {
  this.getRefByUuid((err, ref) => {
    if (err) return callback(err, null)
    this.client.methodCall('VM.snapshot', [this.sessionId, ref, snapshotName], (err, res) => {
      if (err) return callback(err, null)
      this.client.methodCall('VM.get_record', [this.sessionId, res.Value], (err, res) => {
        if (err) return callback(err, null)
        return callback(null, new VM(res.Value, this.client, this.sessionId))
      })
    })
  })
}

VM.prototype.getRefByUuid = function (callback) {
  if (this.ref) return callback(null, this.ref)
  this.client.methodCall('VM.get_by_uuid', [this.sessionId, this.uuid], function (err, res) {
    if (err) return callback(err, null)
    if (res.Value) this.ref = res.Value
    return callback(null, res.Value)
  })
}

VM.prototype.getRawData = function (callback) {
  this.client.methodCall('VM.get_record', [this.sessionId, this.ref], function (err, res) {
    if (err) return callback(err, null)
    return callback(null, res.Value)
  })
}

VM.prototype.remove = function (callback) {
  this.getRefByUuid((err, ref) => {
    if (err) return callback(err, null)
    this.client.methodCall('VM.destroy', [this.sessionId, ref], function (err, res) {
      if (err) return callback(err, null)
      return callback(null, res)
    })
  })
}

module.exports = VM

// PRIVATE METHODS


lib/vdi.js
function VDI (data, client, sessionId) {
  this.uuid = null
  this.name = null
  this.isSnapshot = null
  this.powerState = null
  this.snapshotTime = null
  this.virtualSize = null
  this.physicalUtilisation = null
  this.missing = null
  this.snapshots = []
  this.client = null
  this.sessionId = null
  if (typeof sessionId !== 'undefined') this.sessionId = sessionId
  if (typeof client !== 'undefined') this.client = client
  if (typeof data.uuid !== 'undefined') this.uuid = data.uuid
  if (typeof data.virtual_size !== 'undefined') this.virtualSize = data.virtual_size
  if (typeof data.physical_utilisation !== 'undefined') this.physicalUtilisation = data.physical_utilisation
  if (typeof data.snapshots !== 'undefined') this.snapshots = data.snapshots
  if (typeof data.missing !== 'undefined') this.missing = data.missing
  if (data.name || data.name_label) this.name = data.name || data.name_label
  if (typeof data.is_a_snapshot !== 'undefined') this.isSnapshot = data.is_a_snapshot
  if (typeof data.power_state !== 'undefined') this.powerState = data.power_state.toLowerCase()
  if (typeof data.snapshot_time !== 'undefined') this.snapshotTime = data.snapshot_time
}

VDI.prototype.snapshot = function (snapshotName, callback) {
  this.getRefByUuid((err, ref) => {
    if (err) return callback(err, null)
    this.client.methodCall('VDI.snapshot', [this.sessionId, ref], (err, resSnap) => {
      if (err) return callback(err, null)
      this.client.methodCall('VDI.get_record', [this.sessionId, resSnap.Value], (err, res) => {
        if (err) return callback(err, null)
        return callback(null, new VDI(res.Value, this.client, this.sessionId))
      })
    })
  })
}

VDI.prototype.getRefByUuid = function (callback) {
  if (this.ref) return callback(null, this.ref)
  this.client.methodCall('VDI.get_by_uuid', [this.sessionId, this.uuid], function (err, res) {
    if (err) return callback(err, null)
    if (res.Value) this.ref = res.Value
    return callback(null, res.Value)
  })
}

VDI.prototype.getRawData = function (callback) {
  this.client.methodCall('VDI.get_record', [this.sessionId, this.ref], (err, res) => {
    if (err) return callback(err, null)
    return callback(null, res.Value)
  })
}

VDI.prototype.remove = function (callback) {
  this.getRefByUuid((err, ref) => {
    if (err) return callback(err, null)
    this.client.methodCall('VDI.destroy', [this.sessionId, ref], function (err, res) {
      if (err) return callback(err, null)
      return callback(null, res)
    })
  })
}



module.exports = VDI


lib/xen-module.js
const VM = require('./vm')
const VDI = require('./vdi')
const xmlrpc = require('../node_modules/xmlrpc')
let client = null
let session = null

module.exports = {
  setClient: (host, port, callback) => {
    return setClient(host, port, callback)
  },
  login: (user, password, callback) => {
    return login(user, password, callback)
  },
  getAllVMs: (callback) => {
    return getAllVMs(callback)
  },
  getAllVDIs: (callback) => {
    return getAllVDIs(callback)
  },
  removeVm: (ref, callback) => {
    return removeVm(ref, callback)
  },
  getByUuid: (uuid, callback) => {
    return getByUuid(uuid, callback)
  },
  getByRef: (ref, callback) => {
    return getByRef(ref, callback)
  },
  takeSnapshot: (vm, snapshotName, callback) => {
    return takeSnapshot(vm, snapshotName, callback)
  }
}

function setClient (host, port) {
  client = xmlrpc.createClient({ host: host, port: port })
}

function login (user, password, callback) {
  client.methodCall('session.login_with_password', [user, password], function (err, res) {
    if (err) return callback(err, null)
    session = res.Value
    return callback(null, session)
  })
}

function getAllVMs (callback) {
  let vms = []
  client.methodCall('VM.get_all_records', [session], function (err, res) {
    if (err) return callback(err, null)
    for (let i in res.Value) {
      if (!res.Value[i].is_a_template && !res.Value[i].is_a_snapshot) {
        vms.push(new VM(res.Value[i], client, session))
      }
    }
    return callback(null, vms)
  })
}

function getAllVDIs (callback) {
  let vdis = []
  client.methodCall('VDI.get_all_records', [session], function (err, res) {
    if (err) return callback(err, null)
    for (let i in res.Value) {
      if (!res.Value[i].read_only && !res.Value[i].is_a_snapshot) {
        vdis.push(new VDI(res.Value[i], client, session))
      }
    }
    return callback(null, vdis)
  })
}

function getByUuid (uuid, callback) {
  client.methodCall('VM.get_by_uuid', [session, uuid], function (err, res) {
    if (err) return callback(err, null)
    return callback(null, res)
  })
}

function getByRef (ref, callback) {
  client.methodCall('VM.get_record', [session, ref], function (err, res) {
    if (err) return callback(err, null)
    return callback(null, new VM(res.Value))
  })
}

function takeSnapshot (vm, snapshotName, callback) {
  client.methodCall('VM.snapshot', [session, vm, snapshotName], function (err, res) {
    if (err) return callback(err, null)
    return callback(null, res)
  })
}

function removeVm (ref, callback) {
  client.methodCall('VM.destroy', [session, ref], function (err, res) {
    if (err) return callback(err, null)
    return callback(null, res)
  })
}


cli.js
#! /usr/bin/env node
const program = require('commander')
const moment = require('moment')
const request = require('request')
const fs = require('fs')
const zlib = require('zlib')
const GZip = zlib.createGzip({
  level: 9
})
const XenBackup = require('./')
let connOptions = {
  host: 'localhost',
  port: 80,
  username: 'root',
  password: null,
  sessionId: null
}
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'

program
  .version('0.1.0')
  .option('-H, --host [host]', 'IP or hostname from Xenserver. Default: localhost.')
  .option('-P, --port [port]', 'TCP port from Xenserver. Default: 80.')
  .option('-u, --username [username]', 'Username to login in Xenserver.')
  .option('-p, --password <password>', 'Password to login in Xenserver.')
  .option('-f, --filename <filename>', 'Path to target file of exported snapshot.')
  .option('-t, --type <type>', 'Type of object to export. <vm|vdi> default vm.')
  .option('-v, --vm <vm>', 'VM name label or uuid to backup.')
  .parse(process.argv)

// REQUIRED
let missingRequireds = []
if (!program.type) program.type = 'vm'
if (!program.password) missingRequireds.push('The password of Xenserver user is required as -p or --password option.')
if (!program.vm) missingRequireds.push('The VM is required as -v or --vm option.')
if (!program.filename) missingRequireds.push('The filename is required as -f or --filename option.')
if (missingRequireds.length > 0) {
  for (let i in missingRequireds) {
    console.log(missingRequireds[i])
  }
  process.exit(1)
}

if (program.host) connOptions.host = program.host
if (program.port) connOptions.port = parseInt(program.port)
if (program.username) connOptions.username = program.username
if (program.password) connOptions.password = program.password

XenBackup.setClient(connOptions.host, connOptions.port)
XenBackup.login(connOptions.username, connOptions.password, (err, sessionId) => {
  if (err) {
    console.error(err)
    return process.exit(1)
  }
  connOptions.sessionId = sessionId
  getItens((err, itens) => {
    if (err) {
      console.error(err)
      return process.exit(1)
    }
    if (program.type === 'vm') console.log(`- Searching for VM ${program.vm}`)
    if (program.type === 'vdi') console.log(`- Searching for VDI ${program.vm}`)
    let founded = false
    for (let i in itens) {
      if (itens[i].uuid.toLowerCase() === program.vm.toLowerCase() || itens[i].name.toLowerCase() === program.vm.toLowerCase()) {
        founded = true
        console.log('- Taking snapshot...')
        itens[i].snapshot(itens[i].name.replace(/\s/g, '') + moment().unix(), (err, snapshot) => {
          if (err) {
            console.error(err)
            return process.exit(1)
          }
          console.log(`- Exporting snapshot ${snapshot.name} taked in ${moment(snapshot.snapshotTime).format('LL')}`)
          console.log(`- Will be saved on file ${program.filename}`)
          let protocol = 'http'
          if ((connOptions.port === '443') || (connOptions.port === 443)) protocol = 'https'
          downloadBackup(protocol, snapshot.uuid, (err, download) => {
            if (err) {
              console.error(err)
              return process.exit(1)
            }
            console.log(`- Snapshot exported successfully to ${program.filename}`)
            console.log(`- The file has size ${(fs.statSync(program.filename).size / 1024 / 1024).toFixed(2)}MB`)
            console.log(`- Removing snapshot...`)
            snapshot.remove((err, removed) => {
              if (err) {
                console.error(err)
                return process.exit(1)
              }
              console.log('- Snapshot removed successfuly')
              return process.exit(0)
            })
          })
        })
      }
    }
    if (!founded) {
      console.error('VM or VDI not found in Xenserver.')
      return process.exit(1)
    }
  })
})

function getItens (callback) {
  if (program.type === 'vm') {
    XenBackup.getAllVMs((err, vms) => {
      if (err) return callback(err, null)
      return callback(null, vms)
    })
  }
  if (program.type === 'vdi') {
    XenBackup.getAllVDIs((err, vdis) => {
      if (err) return callback(err, null)
      return callback(null, vdis)
    })
  }
  if ((program.type !== 'vdi') && (program.type !== 'vm')) {
    return callback(new Error('Cannot define object type <vm|vdi>.'))
  }
}

function downloadBackup (protocol, uuid, callback) {
  let sufix = null
  if (program.type === 'vm') {
    sufix = `/export?session_id=${connOptions.sessionId}&use_compression=true&uuid=`
  }
  if (program.type === 'vdi') {
    sufix = `/export_raw_vdi?session_id=${connOptions.sessionId}&format=vhd&vdi=`
  }
  let req = request.get(`${protocol}://${connOptions.host}:${connOptions.port}${sufix}${uuid}`)
    .on('end', function (response) {
      return callback(null, response)
    })
    .on('error', function (err) {
      console.log(err)
      return callback(err, null)
    })
  if (program.type === 'vdi') {
    req.pipe(GZip).pipe(fs.createWriteStream(program.filename))
  } else {
    req.pipe(fs.createWriteStream(program.filename))
  }
}
